### PR TITLE
rgwlc: dispatch S3 notifications on transition

### PR DESCRIPTION
**Pull request by mattbenjamin**
Fixes: https://tracker.ceph.com/issues/58641

Signed-off-by: Matt Benjamin <mbenjamin@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/rgw/rgw_lc.cc [hunk 1]
```diff
@@ -1260,27 +1260,83 @@ class LCOpAction_Transition : public LCOpAction {
     /* If bucket is versioned, create delete_marker for current version
      */
     if (oc.bucket->versioned() && oc.o.is_current() && !oc.o.is_delete_marker()) {
-      ret = remove_expired_obj(oc.dpp, oc, false, rgw::notify::ObjectExpiration);
+      ret = remove_expired_obj(oc.dpp, oc, false, rgw::notify::ObjectTransitionCurrent);
       ldpp_dout(oc.dpp, 20) << "delete_tier_obj Object(key:" << oc.o.key << ") current & not delete_marker" << " versioned_epoch:  " << oc.o.versioned_epoch << "flags: " << oc.o.flags << dendl;
     } else {
-      ret = remove_expired_obj(oc.dpp, oc, true, rgw::notify::ObjectExpiration);
+      ret = remove_expired_obj(oc.dpp, oc, true, rgw::notify::ObjectTransitionNoncurrent);
```
Comment (soumyakoduri) [inline src/rgw/rgw_lc.cc:1266]: else {} case here represents objects' (which can be termed current) transition in the non versioned bucket as well.  So we may need to send different notification event for such objects.

File: src/rgw/rgw_lc.cc [hunk 2]
```diff
@@ -1260,27 +1260,83 @@ class LCOpAction_Transition : public LCOpAction {
...
+    ret = rgw::notify::publish_reserve(oc.dpp, event_type, notify_res, nullptr);
     if (ret < 0) {
+      ldpp_dout(oc.dpp, 1)
+	<< "ERROR: notify reservation failed, deferring transition of object k="
+	<< oc.o.key
+	<< dendl;
       return ret;
     }
 
+    obj = bucket->get_object(oc.o.key);
```
Comment (soumyakoduri) [inline src/rgw/rgw_lc.cc:1324]: Shouldn't this obj be initialized before calling get_notification() above at line:1304?

**Review by soumyakoduri — COMMENTED** [review]
Also these changes seem to address only cloud-transitioned objects. What about regular LC transition?

**Comment by mattbenjamin** [comment]
I don't understand what github is doing here, I did not delete branch linuxbox2/wip-notify-lctrans

**Comment by mattbenjamin** [comment]
ok @cbodley the active PR is https://github.com/ceph/ceph/pull/50680

**Comment by cbodley** [comment]
thanks @mattbenjamin, updated the tracker issue

**Pull request closed without merging**